### ISSUE TITLE
split one nn.linear to two nn.linear, get different output

### ISSUE THREAD

**Issue by NickYangMin** [pytorch/pytorch#106504]
## Describe the bug
when I split one nn.linear to two nn.linear, I get different outputs.

## Code
import torch.nn.functional as F
import torch

batch_size = 1
input_dim = 4096
output_dim = 4096

torch.set_default_tensor_type(torch.cuda.HalfTensor)

input_tensor = torch.randn(batch_size, input_dim)
weight_matrix = torch.randn(output_dim, input_dim)

output_tensor = F.linear(input_tensor, weight_matrix)
print(output_tensor.shape)

output_tensor_half1 = F.linear(input_tensor, weight_matrix[0:output_dim//2,:])
print(output_tensor_half1.shape)
print("output_tensor_half1 diff:", (output_tensor[:,0:output_dim//2] - output_tensor_half1).abs().max())

output_tensor_half2 = F.linear(input_tensor, weight_matrix[output_dim//2:,:])
print(output_tensor_half2.shape)
print("output_tensor_half2 diff:", (output_tensor[:,output_dim//2:] - output_tensor_half2).abs().max())

## result
torch.Size([1, 4096])
torch.Size([1, 2048])
output_tensor_half1 diff: tensor(0.0625)
torch.Size([1, 2048])
output_tensor_half2 diff: tensor(0.1250)

## System Info
Package               Version
--------------------- ------------
accelerate            0.19.0
addict                2.4.0
aiohttp               3.8.4
aiosignal             1.3.1
async-timeout         4.0.2
attrs                 23.1.0
bcrypt                4.0.1
boto3                 1.26.164
botocore              1.29.164
certifi               2023.5.7
cffi                  1.15.1
cfgv                  3.3.1
charset-normalizer    2.1.1
click                 8.1.3
colossalai            0.3.0
contexttimer          0.3.3
contourpy             1.0.7
cpm-kernels           1.0.11
cryptography          41.0.1
cycler                0.11.0
datasets              2.12.0
decorator             5.1.1
dill                  0.3.6
distlib               0.3.6
evaluate              0.4.0
fabric                3.1.0
fairscale             0.4.13
faiss-gpu             1.7.2
filelock              3.12.0
fire                  0.5.0
fonttools             4.40.0
frozenlist            1.3.3
fsspec                2023.5.0
huggingface-hub       0.14.1
identify              2.5.24
idna                  3.4
invoke                2.1.3
jieba                 0.42.1
jmespath              1.0.1
joblib                1.2.0
kiwisolver            1.4.4
markdown-it-py        2.2.0
matplotlib            3.7.1
mdurl                 0.1.2
mmengine              0.8.2
multidict             6.0.4
multiprocess          0.70.14
ninja                 1.11.1
nltk                  3.8
nodeenv               1.8.0
numpy                 1.23.4
openai                0.27.1
opencv-python         4.7.0.72
packaging             23.1
pandas                1.5.3
paramiko              3.2.0
Pillow                9.5.0
pip                   23.1.2
platformdirs          3.5.3
pre-commit            3.3.3
psutil                5.9.5
pyarrow               12.0.0
pycparser             2.21
Pygments              2.15.1
PyNaCl                1.5.0
pyparsing             3.1.0
python-dateutil       2.8.2
pytz                  2023.3
PyYAML                6.0
rank-bm25             0.2.2
regex                 2023.5.5
requests              2.28.1
responses             0.18.0
rich                  13.4.2
s3transfer            0.6.1
safetensors           0.3.1
scikit-learn          1.2.1
scipy                 1.11.1
sentence-transformers 2.2.2
sentencepiece         0.1.99
setuptools            67.8.0
six                   1.16.0
tabulate              0.9.0
termcolor             2.3.0
threadpoolctl         3.1.0
tiktoken              0.4.0
tokenizers            0.13.3
tomli                 2.0.1
torch                 1.13.1+cu116
torchvision           0.14.1+cu116
tqdm                  4.64.1
transformers          4.30.2
typing_extensions     4.7.0
urllib3               1.26.16
virtualenv            20.23.0
wheel                 0.38.4
xxhash                3.2.0
yapf                  0.33.0
yarl                  1.9.2

**Comment by NickYangMin** [comment]
If I want to obtain exactly the same output, what should I do?

**Comment by GilPinsky** [comment]
This is the result of the non-deterministic nature of CUDA operations. What you can do in order to enforce deterministic behavior is setting the CUBLAS_WORKSPACE_CONFIG environment variable at the top of your program as follows:

```python
import os
os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":16:8"
```

**Comment by NickYangMin** [comment]
Thank you for your reply. I try this and get the result:
# result
torch.Size([1, 4096])
torch.Size([1, 2048])
output_tensor_half1 diff: tensor(0.0625)
torch.Size([1, 2048])
output_tensor_half2 diff: tensor(0.1250)

It could be that I expressed myself incorrectly, what I hope to obtain is a diff of 0.

**Comment by GilPinsky** [comment]
Sorry, my bad. The solution I proposed works for torch version 2.0.1. For torch version 1.13.1 that you currently use I could not find a solution that forces deterministic behavior properly.

**Comment by NickYangMin** [comment]
Ok, still very much appreciated.
